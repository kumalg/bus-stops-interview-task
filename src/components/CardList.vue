<template>
  <BaseCard :title="title" :without-paddings="true" class="card-list">
    <div v-if="filterable" class="search-row">
      <SearchInput v-model="userInput" class="search-input" />
    </div>
    <button
      v-if="reversible"
      class="list-title"
      @click="isReversed = !isReversed"
      :tabindex="reversible ? 0 : -1"
      type="button"
    >
      {{ subtitle }}
      <ArrowBottomIcon
        class="order-icon"
        :class="{ flip: isReversed }"
      />
    </button>
    <p v-else class="list-title">
      {{ subtitle }}
    </p>

    
    <transition name="fade" mode="out-in">
      <ul :key="`items_${filteredItems.length}_${isReversed}`" class="list">
        <li
          v-for="item in filteredItems"
          :key="item"
          class="list-item"
        >
          <button :tabindex="selectable ? 0 : -1" class="list-item-button" :class="{ selectable, active: selectable && modelValue === item }" type="button" @click="selectable && selectItem(item)">
            {{ item }}
          </button>
        </li>
      </ul>
    </transition>
  </BaseCard>
</template>

<script setup lang="ts">
import { computed, ref } from 'vue';
import ArrowBottomIcon from '@/icons/ArrowBottomIcon.vue';
import BaseCard from './BaseCard.vue';
import SearchInput from './SearchInput.vue';

const emit = defineEmits<{
  (e: 'update:model-value', item: string): void
}>()

const props = withDefaults(defineProps<{
  items: string[],
  subtitle: string,
  title?: string,
  reversible?: boolean,
  selectable?: boolean,
  filterable?: boolean,
  modelValue?: string | null,
}>(), {
  title: '',
  reversible: false,
  selectable: false,
  modelValue: null
})

const userInput = ref('')
const isReversed = ref(false)

const reversedItem = computed(() => {
  const copy = [...props.items]
  copy.reverse()
  return copy
})

const orderedItems = computed(() => {
  return isReversed.value ? reversedItem.value : props.items
})

const filteredItems = computed(() => {
  if (!props.filterable || !userInput.value) {
    return orderedItems.value
  }

  return orderedItems.value.filter(item => item.toLowerCase().includes(userInput.value.trim().toLowerCase()))
})

const selectItem = (item: string) => {
  emit('update:model-value', item)
}
</script>

<style lang="scss" scoped>
.card-list {
  .list-title {
    display: block;
    background-color: transparent;
    border: none;
    width: 100%;
    font-size: 0.75rem;
    line-height: 1rem;
    letter-spacing: 0.025em;
    font-weight: 600;
    display: flex;
    gap: 0.25rem;
    color: $color-gray-800;
    margin: 0;
    padding: $card-padding;
    border-bottom: 1px solid $color-gray-500;

    &:has(.order-icon) {
      cursor: pointer;

      &:hover .order-icon {
        color: $color-gray-800;
      }
    }

    .order-icon {
      color: $color-gray-700;

      @include default-transition('color, transform');

      &.flip {
        transform: rotate(180deg);
      }
    }
  }

  .search-row {
    padding: 0.5rem;

    .search-input {
      width: 18rem;
      max-width: 100%;
    }
  }

  .list {
    overflow: auto;
    list-style: none;
    padding: 0;
    margin: 0;
    min-height: 0;
  }

  .list-item {
    &:not(:last-child) {
      border-bottom: 1px solid $color-gray-400;
    }
  }

  .list-item-button {
    display: block;
    width: 100%;
    height: 3.5rem;
    border: none;
    background-color: transparent;
    text-align: left;
    padding: 0 $card-padding;
    color: $color-gray-800;
    font-family: inherit;
    user-select: none;

    @include default-transition('color, background-color');

    &.selectable {
      cursor: pointer;

      &:hover, &.hover {
        background-color: $color-gray-300;
      }

      &:active, &.active {
        color: $color-accent;
      }
    }
  }
}
</style>