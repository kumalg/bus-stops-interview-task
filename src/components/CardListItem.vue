<template>
    <li class="card-list-item">
        <button
            :tabindex="selectable ? 0 : -1"
            class="button"
            :class="{ selectable, active: selectable && selectedItem === item }"
            type="button"
            @click="selectable && selectItem()"
        >
            {{ item }}
        </button>
    </li>
</template>

<script setup lang="ts">
const emit = defineEmits<{
    (e: 'select-item', value: string): void;
}>();

const props = defineProps<{
    selectable: boolean;
    selectedItem: string | null;
    item: string;
}>();

const selectItem = () => {
    emit('select-item', props.item);
};
</script>

<style lang="scss" scoped>
.card-list-item {
    .button {
        display: block;
        width: 100%;
        height: 3.5rem;
        border: none;
        background-color: transparent;
        text-align: left;
        padding: 0 $card-padding;
        color: $color-gray-800;
        font-family: inherit;
        user-select: none;

        @include default-transition('color, background-color');

        &.selectable {
            cursor: pointer;

            &:hover,
            &.hover {
                background-color: $color-gray-300;
            }

            &:active,
            &.active {
                color: $color-accent;
            }
        }
    }
}
</style>
