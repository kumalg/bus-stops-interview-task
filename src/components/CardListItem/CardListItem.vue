<template>
    <li class="card-list-item">
        <button
            v-if="selectable"
            class="item selectable"
            tabindex="0"
            type="button"
            :class="{ active: selectable && selectedItem === item }"
            @click="selectable && selectItem()"
        >
            {{ item }}
        </button>
        <div v-else class="item">
            {{ item }}
        </div>
    </li>
</template>

<script setup lang="ts">
const emit = defineEmits<{
    (e: 'select-item', value: string): void;
}>();

const props = defineProps<{
    selectable: boolean;
    selectedItem: string | null;
    item: string;
}>();

const selectItem = () => {
    emit('select-item', props.item);
};
</script>

<style lang="scss" scoped>
.card-list-item {
    .item {
        display: flex;
        align-items: center;
        width: 100%;
        height: 3.5rem;
        border: none;
        background-color: transparent;
        text-align: left;
        padding: 0 $card-padding;
        color: $color-gray-800;
        font-family: inherit;

        @include default-transition('color, background-color');

        &.selectable {
            cursor: pointer;
            user-select: none;

            &:hover,
            &.hover {
                background-color: $color-gray-300;
            }

            &:active,
            &.active {
                color: $color-accent;
            }
        }
    }
}
</style>
