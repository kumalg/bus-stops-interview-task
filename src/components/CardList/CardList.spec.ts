import { flushPromises, mount, VueWrapper } from '@vue/test-utils';

import CardList from './CardList.vue';
import CardListItem from '../CardListItem/CardListItem.vue';

const EXAMPLE_ITEMS = () => ['firstValue', 'secondValue', 'thirdValue', 'differentName'];
const EXAMPLE_ITEMS_REVERSED = () => ['differentName', 'thirdValue', 'secondValue', 'firstValue'];

const EXAMPLE_USER_INPUT = 'val';
const EXAMPLE_ITEMS_FILTERED = () => ['firstValue', 'secondValue', 'thirdValue'];
const EXAMPLE_ITEMS_FILTERED_REVERSED = () => ['thirdValue', 'secondValue', 'firstValue'];

const EXAMPLE_SUBTITLE = 'EXAMPLE_SUBTITLE';
const NO_ITEMS_FOUND_MESSAGE = 'No items found';

describe('CardList', () => {
    const SearchInputStup = {
        template: '<div search-input-stub />'
    };
    const ArrowBottomIconStub = {
        template: '<svg arrow-bottom-icon-stub />'
    };

    const CardListFactory = (props: Partial<InstanceType<typeof CardList>['$props']> = {}) => {
        return mount(CardList, {
            props: {
                items: EXAMPLE_ITEMS(),
                subtitle: EXAMPLE_SUBTITLE,
                ...props
            },
            global: {
                stubs: {
                    SearchInput: SearchInputStup,
                    ArrowBottomIcon: ArrowBottomIconStub
                }
            }
        });
    };

    it('renders proper html markup with only required props', () => {
        const wrapper = CardListFactory({});

        expect(wrapper.element).toMatchSnapshot();
    });

    it('renders proper html markup with opional props', () => {
        const wrapper = CardListFactory({
            filterable: true,
            reversible: true,
            selectable: true
        });

        expect(wrapper.element).toMatchSnapshot();
    });

    describe('search input', () => {
        it('hide if filterable is false', async () => {
            const wrapper = CardListFactory();

            expect(wrapper.findComponent(SearchInputStup).exists()).toBe(false);

            await wrapper.setProps({ filterable: false });

            expect(wrapper.findComponent(SearchInputStup).exists()).toBe(false);
        });

        it('show if filterable is true', () => {
            const wrapper = CardListFactory({
                filterable: true
            });

            expect(wrapper.findComponent(SearchInputStup).exists()).toBe(true);
        });
    });

    it('list title contains subtitle text', () => {
        const wrapper = CardListFactory({
            reversible: false
        });

        expect(wrapper.find('p.list-title').text()).toBe(EXAMPLE_SUBTITLE);
    });

    describe('reverse button', () => {
        it('hide if reversible is false', async () => {
            const wrapper = CardListFactory();

            expect(wrapper.find('button.list-title').exists()).toBe(false);
            expect(wrapper.find('p.list-title').exists()).toBe(true);

            await wrapper.setProps({ reversible: false });

            expect(wrapper.find('button.list-title').exists()).toBe(false);
            expect(wrapper.find('p.list-title').exists()).toBe(true);
        });

        it('show if reversible is true', () => {
            const wrapper = CardListFactory({
                reversible: true
            });

            expect(wrapper.find('button.list-title').exists()).toBe(true);
            expect(wrapper.find('p.list-title').exists()).toBe(false);
        });

        it('contains subtitle text', () => {
            const wrapper = CardListFactory({
                reversible: true
            });

            expect(wrapper.find('button.list-title').text()).toBe(EXAMPLE_SUBTITLE);
        });

        describe('after click', () => {
            it('should reverse arrow icon', async () => {
                const wrapper = CardListFactory({
                    reversible: true
                });

                const iconComponent = wrapper.findComponent(ArrowBottomIconStub);

                expect(iconComponent.classes()).not.toContain('flip');

                await wrapper.find('button.list-title').trigger('click');

                expect(iconComponent.classes()).toContain('flip');
            });
        });
    });

    describe('list items', () => {
        it('should reverse on reverse button click', async () => {
            const wrapper = CardListFactory({
                reversible: true
            });

            expect(wrapper.findAllComponents(CardListItem).map((item) => item.text())).toEqual(
                EXAMPLE_ITEMS()
            );

            await wrapper.find('button.list-title').trigger('click');

            expect(wrapper.findAllComponents(CardListItem).map((item) => item.text())).toEqual(
                EXAMPLE_ITEMS_REVERSED()
            );
        });

        it('should filter items on user input', async () => {
            const wrapper = CardListFactory({
                filterable: true
            });

            expect(wrapper.findAllComponents(CardListItem).map((item) => item.text())).toEqual(
                EXAMPLE_ITEMS()
            );

            wrapper
                .findComponent(SearchInputStup)
                .vm.$emit('update:model-value', EXAMPLE_USER_INPUT);

            await flushPromises();

            expect(wrapper.findAllComponents(CardListItem).map((item) => item.text())).toEqual(
                EXAMPLE_ITEMS_FILTERED()
            );
        });

        it('show correct items when filtered and reversed', async () => {
            const wrapper = CardListFactory({
                filterable: true,
                reversible: true
            });

            expect(wrapper.findAllComponents(CardListItem).map((item) => item.text())).toEqual(
                EXAMPLE_ITEMS()
            );

            await wrapper.find('button.list-title').trigger('click');

            wrapper
                .findComponent(SearchInputStup)
                .vm.$emit('update:model-value', EXAMPLE_USER_INPUT);

            await flushPromises();

            expect(wrapper.findAllComponents(CardListItem).map((item) => item.text())).toEqual(
                EXAMPLE_ITEMS_FILTERED_REVERSED()
            );
        });
    });

    describe('default message', () => {
        const getList = (wrapper: VueWrapper) => {
            return wrapper.find('ul.list');
        };
        const getMessage = (wrapper: VueWrapper) => {
            return wrapper.find('p.message');
        };

        it('hide if no items are passed', () => {
            const wrapper = CardListFactory({
                items: []
            });

            const messageEl = getMessage(wrapper as VueWrapper);

            expect(getList(wrapper as VueWrapper).exists()).toBe(false);
            expect(messageEl.exists()).toBe(false);
        });

        it('hide if items are passed but no items match user input', async () => {
            const wrapper = CardListFactory({
                filterable: true
            });

            wrapper
                .findComponent(SearchInputStup)
                .vm.$emit('update:model-value', 'not existed value');

            await flushPromises();

            const messageEl = getMessage(wrapper as VueWrapper);

            expect(getList(wrapper as VueWrapper).exists()).toBe(false);
            expect(messageEl.exists()).toBe(true);
            expect(messageEl.text()).toBe(NO_ITEMS_FOUND_MESSAGE);
        });
    });
});
