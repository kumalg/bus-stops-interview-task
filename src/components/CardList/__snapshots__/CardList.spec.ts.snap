// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CardList renders proper html markup with only required props 1`] = `
<div
  class="card no-paddings card-list"
>
  <!--v-if-->
  <div
    class="card-body"
  >
    
    <!--v-if-->
    <p
      class="list-title"
    >
      EXAMPLE_SUBTITLE
    </p>
    <transition-stub
      appear="false"
      css="true"
      mode="out-in"
      name="fade"
      persisted="false"
    >
      <ul
        class="list"
      >
        
        <li
          class="card-list-item list-item"
        >
          <button
            class="button"
            tabindex="-1"
            type="button"
          >
            firstValue
          </button>
        </li>
        <li
          class="card-list-item list-item"
        >
          <button
            class="button"
            tabindex="-1"
            type="button"
          >
            secondValue
          </button>
        </li>
        <li
          class="card-list-item list-item"
        >
          <button
            class="button"
            tabindex="-1"
            type="button"
          >
            thirdValue
          </button>
        </li>
        <li
          class="card-list-item list-item"
        >
          <button
            class="button"
            tabindex="-1"
            type="button"
          >
            differentName
          </button>
        </li>
        
      </ul>
    </transition-stub>
    
  </div>
</div>
`;

exports[`CardList renders proper html markup with opional props 1`] = `
<div
  class="card no-paddings card-list"
>
  <!--v-if-->
  <div
    class="card-body"
  >
    
    <div
      class="search-row"
    >
      <div
        class="search-input"
        modelvalue=""
        search-input-stub=""
      />
    </div>
    <button
      class="list-title has-items"
      tabindex="0"
      type="button"
    >
      EXAMPLE_SUBTITLE 
      <svg
        arrow-bottom-icon-stub=""
        class="order-icon"
      />
    </button>
    <transition-stub
      appear="false"
      css="true"
      mode="out-in"
      name="fade"
      persisted="false"
    >
      <ul
        class="list"
      >
        
        <li
          class="card-list-item list-item"
        >
          <button
            class="button selectable"
            tabindex="0"
            type="button"
          >
            firstValue
          </button>
        </li>
        <li
          class="card-list-item list-item"
        >
          <button
            class="button selectable"
            tabindex="0"
            type="button"
          >
            secondValue
          </button>
        </li>
        <li
          class="card-list-item list-item"
        >
          <button
            class="button selectable"
            tabindex="0"
            type="button"
          >
            thirdValue
          </button>
        </li>
        <li
          class="card-list-item list-item"
        >
          <button
            class="button selectable"
            tabindex="0"
            type="button"
          >
            differentName
          </button>
        </li>
        
      </ul>
    </transition-stub>
    
  </div>
</div>
`;
